File: sdk/typescript/src/websocket.ts
/**
 * Aragora WebSocket Client
 *
 * Provides real-time streaming for debate events, agent messages, and consensus.
 */

import type {
  AragoraConfig,
  WebSocketEvent,
  DebateStartEvent,
  RoundStartEvent,
  AgentMessageEvent,
  CritiqueEvent,
  VoteEvent,
  ConsensusEvent,
  DebateEndEvent,
  SynthesisEvent,
  RevisionEvent,
  PhaseChangeEvent,
  AudienceSuggestionEvent,
  UserVoteEvent,
  WarningEvent,
} from './types';

export type WebSocketState = 'connecting' | 'connected' | 'disconnected' | 'reconnecting';

export interface WebSocketOptions {
  /** Auto-reconnect on disconnect (default: true) */
  autoReconnect?: boolean;
  /** Maximum reconnection attempts (default: 5) */
  maxReconnectAttempts?: number;
  /** Base delay between reconnection attempts in ms (default: 1000) */
  reconnectDelay?: number;
  /** Heartbeat interval in ms (default: 30000) */
  heartbeatInterval?: number;
}

type EventHandler<T> = (event: T) => void;

interface EventHandlers {
  connected: EventHandler<void>[];
  disconnected: EventHandler<{ code: number; reason: string }>[];
  error: EventHandler<Error>[];
  debate_start: EventHandler<DebateStartEvent>[];
  round_start: EventHandler<RoundStartEvent>[];
  agent_message: EventHandler<AgentMessageEvent>[];
  propose: EventHandler<AgentMessageEvent>[];
  critique: EventHandler<CritiqueEvent>[];
  revision: EventHandler<RevisionEvent>[];
  synthesis: EventHandler<SynthesisEvent>[];
  vote: EventHandler<VoteEvent>[];
  consensus: EventHandler<ConsensusEvent>[];
  consensus_reached: EventHandler<ConsensusEvent>[];
  debate_end: EventHandler<DebateEndEvent>[];
  phase_change: EventHandler<PhaseChangeEvent>[];
  audience_suggestion: EventHandler<AudienceSuggestionEvent>[];
  user_vote: EventHandler<UserVoteEvent>[];
  warning: EventHandler<WarningEvent>[];
  message: EventHandler<WebSocketEvent>[];
}

export class AragoraWebSocket {
  private ws: WebSocket | null = null;
  private config: AragoraConfig;
  private options: Required<WebSocketOptions>;
  private state: WebSocketState = 'disconnected';
  private reconnectAttempts = 0;
  private heartbeatTimer: ReturnType<typeof setInterval> | null = null;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;

  private handlers: EventHandlers = {
    connected: [],
    disconnected: [],
    error: [],
    debate_start: [],
    round_start: [],
    agent_message: [],
    propose: [],
    critique: [],
    revision: [],
    synthesis: [],
    vote: [],
    consensus: [],
    consensus_reached: [],
    debate_end: [],
    phase_change: [],
    audience_suggestion: [],
    user_vote: [],
    warning: [],
    message: [],
  };

  constructor(config: AragoraConfig, options: WebSocketOptions = {}) {
    this.config = config;
    this.options = {
      autoReconnect: options.autoReconnect ?? true,
      maxReconnectAttempts: options.maxReconnectAttempts ?? 5,
      reconnectDelay: options.reconnectDelay ?? 1000,
      heartbeatInterval: options.heartbeatInterval ?? 30000,
    };
  }

  /**
   * Get the current connection state.
   */
  getState(): WebSocketState {
    return this.state;
  }

  /**
   * Connect to the WebSocket server.
   */
  async connect(debateId?: string): Promise<void> {
    if (this.state === 'connected' || this.state === 'connecting') {
      return;
    }

    this.state = 'connecting';

    const wsUrl = this.buildWsUrl(debateId);

    return new Promise((resolve, reject) => {
      try {
        // Use native WebSocket in browser, or require ws for Node.js
        if (typeof WebSocket !== 'undefined') {
          this.ws = new WebSocket(wsUrl);
        } else {
          // Node.js environment - dynamically require ws
          // eslint-disable-next-line @typescript-eslint/no-var-requires
          const WS = require('ws');
          this.ws = new WS(wsUrl);
        }

        this.ws!.onopen = () => {
          this.state = 'connected';
          this.reconnectAttempts = 0;
          this.startHeartbeat();
          this.emit('connected', undefined);
          resolve();
        };

        this.ws!.onclose = (event) => {
          this.handleDisconnect(event.code, event.reason);
        };

        this.ws!.onerror = (_event) => {
          const error = new Error('WebSocket error');
          this.emit('error', error);
          if (this.state === 'connecting') {
            reject(error);
          }
        };

        this.ws!.onmessage = (event) => {
          this.handleMessage(event.data as string);
        };
      } catch (error) {
        this.state = 'disconnected';
        reject(error);
      }
    });
  }

  /**
   * Disconnect from the WebSocket server.
   */
  disconnect(): void {
    this.options.autoReconnect = false;
    this.cleanup();
    if (this.ws) {
      this.ws.close(1000, 'Client disconnect');
      this.ws = null;
    }
    this.state = 'disconnected';
  }

  /**
   * Subscribe to a specific debate's events.
   */
  subscribe(debateId: string): void {
    this.send({
      type: 'subscribe',
      debate_id: debateId,
    });
  }

  /**
   * Unsubscribe from a debate's events.
   */
  unsubscribe(debateId: string): void {
    this.send({
      type: 'unsubscribe',
      debate_id: debateId,
    });
  }

  /**
   * Send a message to the server.
   */
  send(data: Record<string, unknown>): void {
    if (this.ws && this.state === 'connected') {
      this.ws.send(JSON.stringify(data));
    }
  }

  // Event handlers
  on(event: 'connected', handler: EventHandler<void>): () => void;
  on(event: 'disconnected', handler: EventHandler<{ code: number; reason: string }>): () => void;
  on(event: 'error', handler: EventHandler<Error>): () => void;
  on(event: 'debate_start', handler: EventHandler<DebateStartEvent>): () => void;
  on(event: 'round_start', handler: EventHandler<RoundStartEvent>): () => void;
  on(event: 'agent_message', handler: EventHandler<AgentMessageEvent>): () => void;
  on(event: 'critique', handler: EventHandler<CritiqueEvent>): () => void;
  on(event: 'vote', handler: EventHandler<VoteEvent>): () => void;
  on(event: 'consensus', handler: EventHandler<ConsensusEvent>): () => void;
  on(event: 'debate_end', handler: EventHandler<DebateEndEvent>): () => void;
  on(event: 'message', handler: EventHandler<WebSocketEvent>): () => void;
  on<K extends keyof EventHandlers>(event: K, handler: EventHandlers[K][number]): () => void {
    // TypeScript can't infer indexed mapped type assignments - cast to unknown[] is safe here
    (this.handlers[event] as unknown[]).push(handler);
    return () => this.off(event, handler);
  }

  /**
   * Remove an event handler.
   */
  off<K extends keyof EventHandlers>(event: K, handler: EventHandlers[K][number]): void {
    // TypeScript can't infer indexed mapped type assignments - cast to unknown[] is safe here
    const handlers = this.handlers[event] as unknown[];
    const index = handlers.indexOf(handler);
    if (index !== -1) {
      handlers.splice(index, 1);
    }
  }

  /**
   * Wait for a specific event to occur.
   * Returns a promise that resolves with the event data.
   */
  once<K extends keyof EventHandlers>(event: K): Promise<Parameters<EventHandlers[K][number]>[0]> {
    type EventData = Parameters<EventHandlers[K][number]>[0];
    return new Promise((resolve) => {
      // Handler that removes itself after first call
      const handler = (data: EventData) => {
        // Remove handler - use unknown[] for indexed type compatibility
        (this.handlers[event] as unknown[]).splice(
          (this.handlers[event] as unknown[]).indexOf(handler),
          1
        );
        resolve(data);
      };
      // Add handler - use unknown[] for indexed type compatibility
      (this.handlers[event] as unknown[]).push(handler);
    });
  }

  private buildWsUrl(debateId?: string): string {
    let wsUrl = this.config.wsUrl;

    if (!wsUrl) {
      // Convert HTTP URL to WebSocket URL
      wsUrl = this.config.baseUrl
        .replace(/^http:/, 'ws:')
        .replace(/^https:/, 'wss:');
    }

    // Ensure path ends with /ws
    if (!wsUrl.endsWith('/ws')) {
      wsUrl = wsUrl.replace(/\/?$/, '/ws');
    }

    // Add debate ID if provided
    if (debateId) {
      wsUrl += `?debate_id=${encodeURIComponent(debateId)}`;
    }

    // Add auth token if available
    if (this.config.apiKey) {
      const separator = debateId ? '&' : '?';
      wsUrl += `${separator}token=${encodeURIComponent(this.config.apiKey)}`;
    }

    return wsUrl;
  }

  private handleMessage(data: string): void {
    try {
      const event = JSON.parse(data) as WebSocketEvent;

      // Emit to generic message handlers
      this.emit('message', event);

      // Emit to specific event type handlers - event.data type varies by event type
      const eventType = event.type as keyof EventHandlers;
      if (eventType in this.handlers) {
        // Cast to unknown is safe - emit() will pass to correctly-typed handlers
        this.emit(eventType, event.data as Parameters<EventHandlers[typeof eventType][number]>[0]);
      }
    } catch (error) {
      this.emit('error', new Error(`Failed to parse message: ${data}`));
    }
  }

  private handleDisconnect(code: number, reason: string): void {
    this.cleanup();
    this.state = 'disconnected';
    this.emit('disconnected', { code, reason });

    if (this.options.autoReconnect && this.reconnectAttempts < this.options.maxReconnectAttempts) {
      this.scheduleReconnect();
    }
  }

  private scheduleReconnect(): void {
    this.state = 'reconnecting';
    const delay = this.options.reconnectDelay * Math.pow(2, this.reconnectAttempts);
    this.reconnectAttempts++;

    this.reconnectTimer = setTimeout(() => {
      this.connect().catch(() => {
        // Reconnect failed, will be handled by onclose
      });
    }, delay);
  }

  private startHeartbeat(): void {
    this.heartbeatTimer = setInterval(() => {
      this.send({ type: 'ping' });
    }, this.options.heartbeatInterval);
  }

  private cleanup(): void {
    if (this.heartbeatTimer) {
      clearInterval(this.heartbeatTimer);
      this.heartbeatTimer = null;
    }
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
  }

  private emit<K extends keyof EventHandlers>(
    event: K,
    data: Parameters<EventHandlers[K][number]>[0]
  ): void {
    for (const handler of this.handlers[event]) {
      try {
        (handler as (data: unknown) => void)(data);
      } catch (error) {
        console.error(`Error in ${event} handler:`, error);
      }
    }
  }
}

/**
 * Create a WebSocket client for streaming debate events.
 */
export function createWebSocket(
  config: AragoraConfig,
  options?: WebSocketOptions
): AragoraWebSocket {
  return new AragoraWebSocket(config, options);
}

// =============================================================================
// Async Iterator Support
// =============================================================================

export interface StreamOptions extends WebSocketOptions {
  /** Filter events to only this debate ID */
  debateId?: string;
}

/**
 * Helper to extract loop/debate ID from an event.
 */
function getEventDebateId(event: WebSocketEvent): string | undefined {
  const data = event.data as Record<string, unknown> | undefined;
  const loopId = 'loop_id' in event ? (event as { loop_id?: string }).loop_id : undefined;
  return (
    event.debate_id ||
    loopId ||
    (typeof data?.debate_id === 'string' ? data.debate_id : undefined) ||
    (typeof data?.loop_id === 'string' ? data.loop_id : undefined)
  );
}

/**
 * Create an async iterable stream for debate events.
 *
 * This provides a convenient way to consume debate events using async iteration.
 * The stream will automatically handle WebSocket connection, reconnection, and cleanup.
 *
 * @param config - Aragora configuration (baseUrl, apiKey, etc.)
 * @param options - Stream options including optional debateId filter
 * @returns AsyncGenerator that yields WebSocket events
 *
 * @example
 * ```typescript
 * import { streamDebate } from '@aragora/sdk';
 *
 * const config = { baseUrl: 'http://localhost:8080' };
 * const stream = streamDebate(config, { debateId: 'my-debate-id' });
 *
 * for await (const event of stream) {
 *   console.log(`${event.type}:`, event.data);
 *
 *   if (event.type === 'debate_end') {
 *     break;
 *   }
 * }
 * ```
 *
 * @example
 * ```typescript
 * // Stream all events (no filter)
 * const stream = streamDebate({ baseUrl: 'http://localhost:8080' });
 *
 * for await (const event of stream) {
 *   console.log(event);
 * }
 * ```
 */
export async function* streamDebate(
  config: AragoraConfig,
  options: StreamOptions = {}
): AsyncGenerator<WebSocketEvent, void, unknown> {
  const ws = new AragoraWebSocket(config, options);
  const eventQueue: WebSocketEvent[] = [];
  let resolveNext: ((event: WebSocketEvent) => void) | null = null;
  let rejectNext: ((error: Error) => void) | null = null;
  let ended = false;
  const { debateId } = options;

  // Filter events by debate ID if specified
  const shouldEmit = (event: WebSocketEvent): boolean => {
    if (!debateId) {
      return true;
    }
    const eventDebateId = getEventDebateId(event);
    return !eventDebateId || eventDebateId === debateId;
  };

  // Handle incoming messages
  const unsubMessage = ws.on('message', (event) => {
    if (!shouldEmit(event)) {
      return;
    }

    if (resolveNext) {
      resolveNext(event);
      resolveNext = null;
    } else {
      eventQueue.push(event);
    }

    // Check for terminal events
    if (event.type === 'debate_end' || event.type === 'error') {
      ended = true;
    }
  });

  // Handle errors
  const unsubError = ws.on('error', (error) => {
    if (rejectNext) {
      rejectNext(error);
      rejectNext = null;
    }
    ended = true;
  });

  // Handle disconnection
  const unsubDisconnect = ws.on('disconnected', ({ code, reason }) => {
    ended = true;
    if (resolveNext) {
      // Create a synthetic end event
      resolveNext({
        type: 'debate_end',
        debate_id: debateId,
        timestamp: new Date().toISOString(),
        data: { reason: 'connection_closed', code, close_reason: reason },
      });
      resolveNext = null;
    }
  });

  // Connect to the WebSocket
  await ws.connect(debateId);

  try {
    while (!ended) {
      if (eventQueue.length > 0) {
        yield eventQueue.shift()!;
      } else {
        const event = await new Promise<WebSocketEvent>((resolve, reject) => {
          resolveNext = resolve;
          rejectNext = reject;
        });
        yield event;
      }
    }
  } finally {
    // Cleanup
    unsubMessage();
    unsubError();
    unsubDisconnect();
    ws.disconnect();
  }
}

/**
 * Create a stream that automatically connects and yields events for a specific debate.
 *
 * This is a convenience wrapper around streamDebate that connects to a specific debate.
 *
 * @param config - Aragora configuration
 * @param debateId - The debate ID to stream events for
 * @param options - Additional WebSocket options
 * @returns AsyncGenerator yielding events for the specified debate
 *
 * @example
 * ```typescript
 * const config = { baseUrl: 'http://localhost:8080' };
 * const stream = streamDebateById(config, 'debate-123');
 *
 * for await (const event of stream) {
 *   if (event.type === 'agent_message') {
 *     console.log(`${event.data.agent}: ${event.data.content}`);
 *   }
 * }
 * ```
 */
export function streamDebateById(
  config: AragoraConfig,
  debateId: string,
  options?: WebSocketOptions
): AsyncGenerator<WebSocketEvent, void, unknown> {
  return streamDebate(config, { ...options, debateId });
}

File: sdk/python/aragora/websocket.py
"""
Aragora WebSocket Client

Provides real-time streaming for debate events, agent messages, and consensus.

Usage::

    from sdk.python.aragora.websocket import AragoraWebSocket, WebSocketOptions

    ws = AragoraWebSocket("http://localhost:8080", api_key="your-key")
    ws.on("agent_message", lambda event: print(event))
    await ws.connect(debate_id="debate-123")
"""

from __future__ import annotations

import asyncio
import json
import logging
import queue
from collections.abc import AsyncIterator, Callable
from dataclasses import dataclass, field
from typing import Any
from urllib.parse import quote

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Event types supported by the Aragora WebSocket protocol
# ---------------------------------------------------------------------------

EVENT_TYPES: tuple[str, ...] = (
    "connected",
    "disconnected",
    "error",
    "debate_start",
    "round_start",
    "agent_message",
    "propose",
    "critique",
    "revision",
    "synthesis",
    "vote",
    "consensus",
    "consensus_reached",
    "debate_end",
    "phase_change",
    "audience_suggestion",
    "user_vote",
    "warning",
    "message",
)

# ---------------------------------------------------------------------------
# Data classes
# ---------------------------------------------------------------------------


@dataclass
class WebSocketOptions:
    """Configuration options for the WebSocket connection."""

    auto_reconnect: bool = True
    max_reconnect_attempts: int = 5
    reconnect_delay: float = 1.0
    heartbeat_interval: float = 30.0


@dataclass
class WebSocketEvent:
    """A single event received over the WebSocket connection.

    Attributes:
        type: Event type string (e.g. ``"debate_start"``, ``"agent_message"``).
        data: Raw event payload as a dictionary.
        timestamp: ISO timestamp of when the event occurred.
        debate_id: ID of the debate this event belongs to.
        typed_data: Parsed typed event object, if available. Check this for
            type-safe access to event fields instead of using ``data``.
    """

    type: str
    data: dict[str, Any] = field(default_factory=dict)
    timestamp: str = ""
    debate_id: str | None = None
    typed_data: Any = None


# ---------------------------------------------------------------------------
# WebSocket state constants
# ---------------------------------------------------------------------------

_STATE_CONNECTING = "connecting"
_STATE_CONNECTED = "connected"
_STATE_DISCONNECTED = "disconnected"
_STATE_RECONNECTING = "reconnecting"


# ---------------------------------------------------------------------------
# AragoraWebSocket
# ---------------------------------------------------------------------------


class AragoraWebSocket:
    """Real-time WebSocket client for streaming Aragora debate events.

    Parameters:
        base_url: The HTTP base URL of the Aragora server (e.g. ``http://localhost:8080``).
        api_key:  Optional API key used for authentication.
        ws_url:   Explicit WebSocket URL.  When *None*, derived from *base_url*.
        options:  Connection tuning knobs (reconnect behaviour, heartbeat, etc.).
    """

    def __init__(
        self,
        base_url: str,
        api_key: str | None = None,
        ws_url: str | None = None,
        options: WebSocketOptions | None = None,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.ws_url = ws_url
        self.options = options or WebSocketOptions()

        # Internal state
        self._state: str = _STATE_DISCONNECTED
        self._ws: Any = None  # websockets connection object
        self._reconnect_attempts: int = 0
        self._heartbeat_task: asyncio.Task[None] | None = None
        self._receive_task: asyncio.Task[None] | None = None
        self._reconnect_task: asyncio.Task[None] | None = None
        self._closing: bool = False

        # Event handler registry: event_type -> list[handler]
        self._handlers: dict[str, list[Callable[..., Any]]] = {evt: [] for evt in EVENT_TYPES}

        # Queue used for the async-iterator interface (``events()``).
        # We use ``queue.Queue`` (thread-safe, sync put) so that
        # ``_handle_message`` can enqueue without needing ``await``.
        self._event_queue: queue.Queue[WebSocketEvent | None] = queue.Queue()

        # Debate subscriptions tracked locally
        self._subscriptions: set[str] = set()

    # -- public properties ---------------------------------------------------

    @property
    def state(self) -> str:
        """Return the current connection state string."""
        return self._state

    # -- connection lifecycle ------------------------------------------------

    async def connect(self, debate_id: str | None = None) -> None:
        """Open a WebSocket connection to the server.

        If *debate_id* is given it is included as a query parameter so that the
        server immediately scopes the stream to that debate.
        """
        if self._state in (_STATE_CONNECTED, _STATE_CONNECTING):
            return

        self._state = _STATE_CONNECTING
        self._closing = False
        url = self._build_ws_url(debate_id)

        try:
            # Import websockets lazily so the module can be loaded even when
            # the optional dependency is not installed.
            import websockets  # type: ignore[import-untyped]

            self._ws = await websockets.connect(url)  # type: ignore[attr-defined]
            self._state = _STATE_CONNECTED
            self._reconnect_attempts = 0

            # Start background tasks
            self._receive_task = asyncio.ensure_future(self._receive_loop())
            self._heartbeat_task = asyncio.ensure_future(self._heartbeat_loop())

            self._emit("connected", {})
        except Exception as exc:
            self._state = _STATE_DISCONNECTED
            self._emit("error", {"error": str(exc)})
            raise

    async def close(self) -> None:
        """Gracefully close the WebSocket connection."""
        self._closing = True
        self.options.auto_reconnect = False
        self._cleanup()

        if self._ws is not None:
            try:
                await self._ws.close()
            except Exception:
                pass
            self._ws = None

        self._state = _STATE_DISCONNECTED

        # Signal the async-iterator that we are done.
        self._event_queue.put(None)

    # -- subscriptions -------------------------------------------------------

    def subscribe(self, debate_id: str) -> None:
        """Subscribe to events for a specific debate."""
        self._subscriptions.add(debate_id)
        self._send({"type": "subscribe", "debate_id": debate_id})

    def unsubscribe(self, debate_id: str) -> None:
        """Unsubscribe from a debate's event stream."""
        self._subscriptions.discard(debate_id)
        self._send({"type": "unsubscribe", "debate_id": debate_id})

    # -- event handler registration ------------------------------------------

    def on(self, event: str, handler: Callable[..., Any]) -> Callable[[], None]:
        """Register *handler* to be called whenever *event* fires.

        Returns a callable that, when invoked, removes the handler (i.e. an
        *unsubscribe* function).
        """
        if event not in self._handlers:
            self._handlers[event] = []
        self._handlers[event].append(handler)

        def _unsubscribe() -> None:
            self.off(event, handler)

        return _unsubscribe

    def off(self, event: str, handler: Callable[..., Any]) -> None:
        """Remove a previously registered handler for *event*."""
        handlers = self._handlers.get(event)
        if handlers is not None:
            try:
                handlers.remove(handler)
            except ValueError:
                pass

    # -- async iterator interface --------------------------------------------

    async def events(self) -> AsyncIterator[WebSocketEvent]:
        """Yield events as an async iterator.

        The iterator terminates when the connection is closed.

        Example::

            async for event in ws.events():
                print(event.type, event.data)
        """
        loop = asyncio.get_event_loop()
        while True:
            # Retrieve from the sync queue without blocking the event loop.
            event = await loop.run_in_executor(None, self._event_queue.get)
            if event is None:
                # Sentinel: stream finished.
                break
            yield event

    # -- internals -----------------------------------------------------------

    def _build_ws_url(self, debate_id: str | None = None) -> str:
        """Construct the full WebSocket URL including auth and debate params."""
        ws_url = self.ws_url

        if not ws_url:
            ws_url = self.base_url.replace("http://", "ws://").replace("https://", "wss://")

        # Ensure the path ends with /ws
        if not ws_url.endswith("/ws"):
            ws_url = ws_url.rstrip("/") + "/ws"

        # Build query string
        params: list[str] = []
        if debate_id:
            params.append(f"debate_id={quote(debate_id, safe='')}")
        if self.api_key:
            params.append(f"token={quote(self.api_key, safe='')}")

        if params:
            ws_url += "?" + "&".join(params)

        return ws_url

    def _handle_message(self, raw: str) -> None:
        """Parse and dispatch an incoming message."""
        try:
            payload = json.loads(raw)
        except (json.JSONDecodeError, TypeError):
            self._emit("error", {"error": f"Failed to parse message: {raw}"})
            return

        event = WebSocketEvent(
            type=payload.get("type", "message"),
            data=payload.get("data", {}),
            timestamp=payload.get("timestamp", ""),
            debate_id=payload.get("debate_id"),
        )

        # Parse typed event data if a matching event class exists
        from .events import EVENT_CLASS_MAP

        cls = EVENT_CLASS_MAP.get(event.type)
        if cls:
            try:
                # Only pass fields that exist in the dataclass
                event.typed_data = cls(
                    **{k: v for k, v in event.data.items() if k in cls.__dataclass_fields__}
                )
            except (TypeError, KeyError):
                pass  # Fall back to untyped data dict

        # Emit to the generic ``message`` handlers.
        self._emit("message", event)

        # Emit to specific event-type handlers.
        if event.type in self._handlers:
            self._emit(event.type, event)

        # Also enqueue for the async iterator.
        self._event_queue.put(event)

    def _emit(self, event: str, data: Any) -> None:
        """Invoke all registered handlers for *event*."""
        for handler in list(self._handlers.get(event, [])):
            try:
                handler(data)
            except Exception:
                logger.exception("Error in %s handler", event)

    def _send(self, data: dict[str, Any]) -> None:
        """Queue a JSON message for sending (fire-and-forget)."""
        if self._ws is not None and self._state == _STATE_CONNECTED:
            try:
                asyncio.ensure_future(self._ws.send(json.dumps(data)))
            except Exception:
                logger.debug("Failed to send message", exc_info=True)

    # -- background loops ----------------------------------------------------

    async def _receive_loop(self) -> None:
        """Continuously read messages from the WebSocket."""
        try:
            async for raw_message in self._ws:
                self._handle_message(
                    raw_message if isinstance(raw_message, str) else raw_message.decode()
                )
        except Exception as exc:
            if not self._closing:
                self._emit("error", {"error": str(exc)})
                self._handle_disconnect(1006, str(exc))

    async def _heartbeat_loop(self) -> None:
        """Periodically send a ping frame to keep the connection alive."""
        try:
            while self._state == _STATE_CONNECTED:
                await asyncio.sleep(self.options.heartbeat_interval)
                if self._state == _STATE_CONNECTED:
                    self._send({"type": "ping"})
        except asyncio.CancelledError:
            pass

    # -- reconnection --------------------------------------------------------

    def _handle_disconnect(self, code: int, reason: str) -> None:
        """Handle an unexpected disconnect."""
        self._cleanup()
        self._state = _STATE_DISCONNECTED
        self._emit("disconnected", {"code": code, "reason": reason})

        if (
            self.options.auto_reconnect
            and self._reconnect_attempts < self.options.max_reconnect_attempts
        ):
            self._schedule_reconnect()

    def _schedule_reconnect(self) -> None:
        """Schedule a reconnection with exponential back-off."""
        self._state = _STATE_RECONNECTING
        delay = self.options.reconnect_delay * (2**self._reconnect_attempts)
        self._reconnect_attempts += 1

        async def _do_reconnect() -> None:
            await asyncio.sleep(delay)
            try:
                await self.connect()
            except Exception:
                pass  # connect() will emit errors and trigger disconnect

        self._reconnect_task = asyncio.ensure_future(_do_reconnect())

    def _cleanup(self) -> None:
        """Cancel background tasks."""
        for task in (self._heartbeat_task, self._receive_task, self._reconnect_task):
            if task is not None and not task.done():
                task.cancel()
        self._heartbeat_task = None
        self._receive_task = None
        self._reconnect_task = None


# ---------------------------------------------------------------------------
# Convenience: async generator for streaming debate events
# ---------------------------------------------------------------------------


async def stream_debate(
    base_url: str,
    debate_id: str | None = None,
    api_key: str | None = None,
    options: WebSocketOptions | None = None,
) -> AsyncIterator[WebSocketEvent]:
    """High-level async generator that connects, streams, and cleans up.

    Example::

        async for event in stream_debate("http://localhost:8080", debate_id="d1"):
            print(event.type, event.data)
            if event.type == "debate_end":
                break
    """
    ws = AragoraWebSocket(base_url, api_key=api_key, options=options)
    await ws.connect(debate_id=debate_id)

    try:
        async for event in ws.events():
            yield event
            if event.type in ("debate_end", "error"):
                break
    finally:
        await ws.close()
